python/cudf/cudf/pylibcudf_tests/test_copying.py

```diff
@@ -0,0 +1,579 @@
+    for i in range(target_table.num_rows()):
+        if (source_row := pa.compute.index_in(i, pa_input_column)).is_valid:
+            output_rows.append(pa_source_table.take([source_row]))
+        else:
+            output_rows.append(pa_target_table.take([i]))
+
+    expected = pa.concat_tables(output_rows)
+    assert_table_eq(result, expected)
+
+
+def test_scatter_table_num_col_mismatch(
+    source_table, input_column, target_table
+):
+    # Number of columns in source and target must match.
+    with pytest.raises(ValueError):
+        plc.copying.scatter(
+            plc.Table(source_table.columns()[:2]),
+            input_column,
+            target_table,
+        )
+
+
+def test_scatter_table_num_row_mismatch(source_table, target_table):
+    # Number of rows in source and scatter map must match.
+    with pytest.raises(ValueError):
+        plc.copying.scatter(
+            source_table,
+            column_from_arrow(pa.array(range(source_table.num_rows() * 2))),
+            target_table,
+        )
+
+
+def test_scatter_table_map_has_nulls(source_table, target_table):
+    with pytest.raises(ValueError):
+        plc.copying.scatter(
+            source_table,
+            column_from_arrow(pa.array([None] * source_table.num_rows())),
+            target_table,
+        )
+
+
+def test_scatter_table_type_mismatch(source_table, input_column, target_table):
+    with pytest.raises(TypeError):
+        pa_array = pa.array([True] * source_table.num_rows())
+        ncol = source_table.num_columns()
+        pa_table = pa.table([pa_array] * ncol, [""] * ncol)
+        plc.copying.scatter(
+            plc.Table.from_arrow(pa_table),
+            input_column,
+            target_table,
+        )
+
+
+def test_scatter_scalars(source_scalar, input_column, target_table):
+    result = plc.copying.scatter(
+        [source_scalar] * target_table.num_columns(),
+        input_column,
+        target_table,
+    )
+    # TODO: There's no reason to require column metadata in this API. We can construct a
+    # placeholder internally and discard it after the conversion if necessary.
+    host_scalar = scalar_to_arrow(source_scalar).as_py()
+    pa_input_column = column_to_arrow(input_column)
+    pa_target_table = table_to_arrow(target_table)
+
+    # TODO: Is there a cleaner vectorized way to do this in pyarrow?
+    arrays = []
+    for i in range(target_table.num_columns()):
+        values = []
+        for j in range(target_table.num_rows()):
+            if pa.compute.is_in(j, pa_input_column).as_py():
+                values.append(host_scalar)
+            else:
+                values.append(pa_target_table.column(i)[j].as_py())
+        arrays.append(pa.array(values))
+
+    expected = pa.table(arrays, [""] * target_table.num_columns())
+    assert_table_eq(result, expected)
+
+
+def test_scatter_scalars_num_scalars_mismatch(
+    source_scalar, input_column, target_table
+):
+    with pytest.raises(ValueError):
```

Comment:
I think we may consider matching on at least part of the string, even though our wrappers are thin it seems possible something else could throw a ValueError that isn't libcudf

Reply:
The most that I would support doing is matching the prefix "CUDF failure at", which we encode in the `CUDF_EXPECTS` and `CUDF_FAIL` macros. In general we don't want libcudf to have to make any promises as to the stability of the exception messages. I'll add a simple example of what I'd propose and we can go from there.